lib/libxdp/xsk.c

```diff
@@ -306,24 +297,42 @@ static int xsk_create_umem_rings(struct xsk_umem *umem, int fd,
 	return err;
 }
 
-int xsk_umem__create_with_fd(struct xsk_umem **umem_ptr, int fd,
-			     void *umem_area, __u64 size,
-			     struct xsk_ring_prod *fill,
-			     struct xsk_ring_cons *comp,
-			     const struct xsk_umem_config *usr_config)
-{
+struct xsk_umem *xsk_umem__create_opts(void *umem_area,
+				 struct xsk_ring_prod *fill,
+				 struct xsk_ring_cons *comp,
+				 struct xsk_umem_opts *opts) {
 	struct xdp_umem_reg mr;
 	struct xsk_umem *umem;
-	int err;
+	int err, fd;
+	__u64 size;
+	
+	if (!umem_area || !fill || !comp) {
+		err = -EFAULT;
+		goto err;
+	}
 
-	if (!umem_area || !umem_ptr || !fill || !comp)
-		return -EFAULT;
-	if (!size && !xsk_page_aligned(umem_area))
-		return -EINVAL;
+	if (!opts) {
```

Comment:
No need for this branch; `OPTS_GET` and `OPTS_VALID` correctly handles a NULL value of opts, so you can just unconditionally use the code in the `else` branch

Reply:
> Better to have the mandatory fields be direct function arguments, and have the options only be the, well, optional values. That way, a caller can also pass a NULL opts argument if they are happy with the defaults...
Since `OPTS_VALID` will return `false` if `opts` is NULL, there is no way to let a caller to pass a NULL `opts` if `if` branch is removed, maybe we still need this `if`?